INT-3944: Async JMS Outbound Gateway

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3944

garyrussell committed:
INT-3944: Async JMS Outbound Gateway

JIRA: https://jira.spring.io/browse/INT-3944

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -560,6 +578,17 @@ protected void doInit() {
 				container.afterPropertiesSet();
 				this.replyContainer = container;
+				if (this.correlationKey == null) {
+					logger.warn("'async=true' requires a correlationKey; ignored");
```
???
No guaranty that I have set `async=true` before.
Looks like condition has been missed around this `warn`.

garyrussell replied:
Good catch :smile:

artembilan commented on src/reference/asciidoc/jms.adoc:
```diff
@@ -397,6 +397,21 @@ __NOT ALLOWED__
+
+By default, when a request is sent to the gateway, the requesting thread is suspended until the reply is received and
+the flow then continues on that thread.
+If `async` is true, the requesting thread is released immediately and the reply is returned (and the flow continues)
```
`immediately after `send` operation and ...` ?

garyrussell replied:
:+1: 

immediately after the send operation...

garyrussell committed:
Polishing (PR comments) and Doc Polish

garyrussell commented:
Rebased, pushed.

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -457,6 +463,18 @@ public void setIdleReplyContainerTimeout(long idleReplyContainerTimeout, TimeUni
+	 * @param async the async to set.
+	 * @since 4.3
+	 */
+	public void setAsync(boolean async) {
```
I think after merging #1739 we can remove this setter and its property altogether.
Having the `protected getAsyncReplySupported()` we good to go with your internal logic around `this.correlationKey`.

garyrussell replied:
:+1:

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsTimeoutException.java:
```diff
@@ -0,0 +1,33 @@
+ *
+ */
+@SuppressWarnings("serial")
+public class JmsTimeoutException extends RuntimeException {
```
Shouldn't we extend here from the `org.springframework.jms.JmsException`?
And yes, I really don't see there an appropriate exception for us...

artembilan commented:
That's all for today.
I'll look deeply to the `JmsOutboundGateway` refactoring tomorrow.

garyrussell committed:
Polishing - PR Comments

Remove custom `async` boolean in favor of the now public setter.

Add exception to `JmsException` hierarchy.

garyrussell commented:
Rebased; pushed.

artembilan commented on src/reference/asciidoc/jms.adoc:
```diff
@@ -397,6 +397,23 @@ __NOT ALLOWED__
+[[jms-async-gateway]]
+==== Async Gateway
+
+Starting with _version 4.3_, you can now specify `async="true"` (or `setAsync(true)`) when configuring the outbound
```
s/setAsync/setAsyncReplySupported

Or maybe it still isn't late to rename that target property (and its getter/setter) just to the `async` and  that's all?

garyrussell replied:
Yes; I thought about that - shorter :smirk:

artembilan commented on src/reference/asciidoc/jms.adoc:
```diff
@@ -397,6 +397,23 @@ __NOT ALLOWED__
+This can be useful when the gateway is invoked on a poller thread; the thread is released and is available for other
+tasks within the framework.
+
+`async` requires a `<reply-listener/>` (or `setUseReplyContainer(true)` when using Java configuration); it also
```
Maybe "The `async`"?
Otherwise it looks a bit weird to star the sentence from the lower case.

garyrussell replied:
The `async` property requires...

would be ok

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -1112,7 +1152,43 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
+
+	private void expire(String correlationId) {
+		final SettableListenableFuture<Message<?>> future = JmsOutboundGateway.this.futures
+				.remove(correlationId);
```
The `JmsOutboundGateway.` is redundant and the `remove` can be on the same line.
Doesn't hurt, of course. Will fix on merge.

artembilan commented:
Not related, but can be fixed as well. 
There is a typo in JavaDoc:
````java
/**
 * @param replyContainerProperties the replyContainerproperties to set
 */
public void setReplyContainerProperties(ReplyContainerProperties replyContainerProperties) {
````

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -560,6 +564,16 @@ protected void doInit() {
 				container.afterPropertiesSet();
 				this.replyContainer = container;
+				if (isAsyncReplySupported() && this.correlationKey == null) {
+					logger.warn("'async=true' requires a correlationKey; ignored");
```
I wonder how Checkstyle misses `logger` without `this.`...

artembilan replied:
Ah! I know. Just because it doesn't see that the `logger` is from the super class :smile:

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -726,34 +742,44 @@ protected Object handleRequestMessage(final Message<?> message) {
+			replyMessage = this.getMessageBuilderFactory().fromMessage((Message<?>) result).copyHeaders(jmsReplyHeaders).build();
+		}
+		else {
+			replyMessage = this.getMessageBuilderFactory().withPayload(result).copyHeaders(jmsReplyHeaders).build();
```
Would be better to return to the `AbstractMessageProducingHandler` just `MessageBuilder`.
The last one does exactly this:
```
if (reply instanceof Message) {
	builder = this.getMessageBuilderFactory().fromMessage((Message<?>) reply);
}
else if (reply instanceof AbstractIntegrationMessageBuilder) {
	builder = (AbstractIntegrationMessageBuilder<?>) reply;
}
```
So, it might be a bit sad to recreate message all the time.

From other side I see that I use `reply` for the `getOutputChannelFromRoutingSlip()`, so `MessageBuilder` might expose access to its `payload` and `header` on the matter...

But that is a different story and should be addressed in the separate JIRA.
Right now I just want to change this code to return Builder not the final message.

WDYT?

mbogoevici replied:
FWIW, I'll venture and say that cascaded message re-creation to update headers, especially due to timestamp and id generation, created performance bottlenecks while we did performance testing a year ago. it surely fades against serialization and 'real' processing, but it also adds undue overhead, all things considered, so dealing with the MessageBuilder only until the final send would be nice.

artembilan replied:
Thank you Marius for feedback! I fixed that for this JMS case. See my polishing

garyrussell replied:
We introduced the `MutableMessage` explicitly for this purpose (when a message is rebuilt several times within the framework as it is passed back up the call stack) but we never took the next step and actually used it for that purpose.

artembilan committed:
INT-3944: Polishing

* Rename `asyncReplySupported` just to `async`. As well as its getter and setter. Plus fix docs on the matter
* Polishing for the `JmsOutboundGateway` according PR comments
* Rework `JmsOutboundGateway` to return `AbstractIntegrationMessageBuilder` instead of `Message`
since `AbstractMessageProducingHandler` rebuilds `Message` for a new one to copy headers from request
* Add `getPayload()` and `getHeaders()` to the `AbstractIntegrationMessageBuilder` to make the `Routing Slip`
users happy, when the `AbstractIntegrationMessageBuilder` is pushed to the `Routing Slip path` function
* Fix race condition in the `ChatMessageListeningEndpointTests`: the `stanza` parsing is done in the different Thread.

artembilan commented:
Here is a polishing regarding that what I think on the matter: https://github.com/artembilan/spring-integration/commit/26fbcdfc0773244733e13cb90dc6550c7ceb5e21

See the commit comment.

If you find it so complicated, we can consider to merge this your PR and raise a new one for that.

garyrussell committed:
Doc Polishing (Routing Slip)

garyrussell commented:
Polish looks good; thanks, but I added some doc polish for the routing slip.

artembilan commented:
LGTM
Merging...
Will fix timing issue on merge:
```
org.springframework.integration.file.filters.LastModifiedFileListFilterTests > testAge FAILED
    java.lang.AssertionError at LastModifiedFileListFilterTests.java:47
```

artembilan commented:
Merge as https://github.com/spring-projects/spring-integration/commit/4bfcdb9dfaf270b76f8b30eddc3ae64f2bcd48a0
